daemon/snap.go

```diff
@@ -239,7 +239,7 @@ func mapLocal(about aboutSnap, sd clientutil.StatusDecorator) *client.Snap {
 	if !about.gatingHold.IsZero() {
 		result.GatingHold = &about.gatingHold
 	}
-
+	result.RefreshInhibitProceedTime = snapst.RefreshInhibitProceedTime(st)
```

Comment:
we need to see how to lock the state for this later,  locking/unlocking for each snap seems a bit wasteful. Maybe a stranger but maybe better API would be to have something that take state and returns a function

snapstate.RefreshInhibitProceedTime(state *State) func(*snapstate.SnapState) time.Time{}

so we lock and get the paremeter(s) only once and return a closure that uses them

and we pass the function to mapLocal instead of state

Reply:
I think there is something I am not fully understanding in this approach.

why would the function returned from snapstate need snapstate as parameter? how would we know if the state is already locked from inside? where would the locking/unlocking happen?

wouldn't it be simpler if we just assume the state is locked?
```diff
diff --git a/daemon/api_snaps.go b/daemon/api_snaps.go
index 7138e2b791..3a1c9a855c 100644
--- a/daemon/api_snaps.go
+++ b/daemon/api_snaps.go
@@ -90,7 +90,9 @@ func getSnapInfo(c *Command, r *http.Request, user *auth.UserState) Response {
 
        sd := servicestate.NewStatusDecorator(progress.Null)
 
+       st.Lock()
        result := webify(mapLocal(st, about, sd), url.String())
+       st.Unlock()
 
        return SyncResponse(result)
 }
@@ -922,6 +924,8 @@ func getSnapsInfo(c *Command, r *http.Request, user *auth.UserState) Response {
 
        results := make([]*json.RawMessage, len(found))
 
+       st.Lock()
+       defer st.Unlock()
        sd := servicestate.NewStatusDecorator(progress.Null)
        for i, x := range found {
                name := x.info.InstanceName()
diff --git a/daemon/snap.go b/daemon/snap.go
index 63f53cc33d..4979509603 100644
--- a/daemon/snap.go
+++ b/daemon/snap.go
@@ -200,6 +200,7 @@ func clientHealthFromHealthstate(h *healthstate.HealthState) *client.SnapHealth
        }
 }
 
+// The provided state must be locked by the caller.
 func mapLocal(st *state.State, about aboutSnap, sd clientutil.StatusDecorator) *client.Snap {
        localSnap, snapst := about.info, about.snapst
        result, err := clientutil.ClientSnapFromSnapInfo(localSnap, sd)
diff --git a/overlord/snapstate/snapmgr.go b/overlord/snapstate/snapmgr.go
index b2a1e3f0a9..b0de693aed 100644
--- a/overlord/snapstate/snapmgr.go
+++ b/overlord/snapstate/snapmgr.go
@@ -534,6 +534,8 @@ func (snapst *SnapState) InstanceName() string {
 // RefreshInhibitProceedTime is the time after which a pending refresh is forced
 // for a running snap in the next auto-refresh. If nil is returned, then there
 // are no pending refreshes.
+//
+// The provided state must be locked by the caller.
 func (snapst *SnapState) RefreshInhibitProceedTime(st *state.State) *time.Time {
        if snapst.RefreshInhibitedTime == nil {
                return nil
```

Reply:
the code is trying not to be another place where we take the lock

Reply:
I updated the code to get proceedTime values inside `allLocalSnapInfos` and `localSnapInfo` and attach it to `aboutSnap` (which is already done for `hold` and `gatingHold`) to utilize the already held lock. Would that be sufficient?